event_model/__init__.py

```diff
@@ -439,3 +440,113 @@ def _transpose_dict_of_lists(dict_of_lists):
+        read (i.e. use `if verified_filled(event_page) ...`) it will never
+        return False as it will raise an exception.
+    '''
+
+    if not all(map(all, event_page['filled'].values())):
+        # check that all event_page data is filled.
+        unfilled_data = []
+        for field in event_page['filled']:
+            if not event_page['filled'][field]:
+                for field, filled in event_page['filled'].items():
+                    if not all(filled):
+                        unfilled_data.append(field)
+                        # Note: As of this writing, this is a slightly
+                        # aspirational error message, as event_model.Filler has
+                        # not been merged yet. May need to be revisited if it
+                        # is renamed or kept elsewhere in the end.
+                        raise UnfilledData("unfilled data found in "
+                                           "{!r}. Try passing the parameter "
+                                           "`gen` through `event_model.Filler`"
+                                           " first.".format(unfilled_data))
+    else:
+        verified = True
+        return verified
+
+
+def sanitize_doc(doc):
+    '''Takes in an event-model document and returns a copy which has all the
+    numpy objects converted to built-in pyton versions.
+
+    This is useful for sanitzing documents prior to sending to a mongo database
+    or a json file.
+
+    Parameters
+    ----------
+    doc : event-model document.
+        The event-model document to be sanitized
+
+    Returns
+    -------
+    sanitized_doc : event-model document
+        The event-model document with numpy objects converted to built-in pyton
+        types.
+    '''
+    sanitized_doc = doc.copy()
+    _apply_to_dict_recursively(doc, _sanitize_numpy)
+
+    return sanitized_doc
+
+
+def _sanitize_numpy(val):
+    '''Convert any numpy objects into built-in Python types.
+
+    Parameters
+    ----------
+    val : The potential numpy object to be converted.
+
+    Returns
+    -------
+    val : The input parameter, converted to a built-in python type if it is a
+        numpy type.
+    '''
+    if isinstance(val, (numpy.generic, numpy.ndarray)):
+        if numpy.isscalar(val):
+            return val.item()
+        return val.tolist()
+    return val
+
+
+def _apply_to_dict_recursively(dictionary, func):
+    '''
+    step through a dictionary of dictionaries recursively and apply a function
+    to each value.
+
+    Parameters
+    ----------
+    dictionary : dict
+        The dictionary of dictionaries to be recursivly searched.
+    func : function
+        A function to apply to each value in dictionary.
+    '''
+
+    for key, val in dictionary.items():
+        if hasattr(val, 'items'):
+            dictionary[key] = _apply_to_dict_recursively(val, func)
+        dictionary[key] = func(val)
+
+
+class UnfilledData(Exception):
```

Comment:
We have adopted* the convention that every exception _directly_ raised by a library (i.e. typed out as `raise ___`, not caused indirectly) should inherit from some base exception defined in that library. In the case of this library, every error raised by event_model is some subclass of a `EventModelError`. This helps users see which layer a given error is coming from.

Some exceptions are simple variants of Python's built-in errors. For example, instead of raising `KeyError`, we raise `EventModelKeyError` which is a subclass of both `KeyError` and `EventModelError`. Therefore, it can be "caught" via `except KeyError` just like a normal `KeyError` would be, but it also signals that it is coming from event-model.

We can also have special errors that have no built-in counterpart. There is no examples of this in event-model yet, but there are some in caproto, such as `RemoteProtocolError`. It inherits from `CaprotoError`, as even error raised by caproto does, but does not inherit from anything else.

I think this is in that latter category.

*This is a relatively recent (1-2 year) practice for us. We do it fully consistently in caproto, inconsistently in ophyd, and hardly at all yet in bluesky or databroker.

Reply:
I am for this standard, but once again raise the fact that word of mouth is not a useful or adequate way of passing this information to new contributors. For a projects with a large, multi-facility, contributor base we need to have a `standards document` that encompasses these.